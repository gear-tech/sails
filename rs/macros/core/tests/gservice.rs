use proc_macro2::TokenStream;
use quote::quote;
use sails_macros_core::__gservice_internal as gservice;

#[test]
fn works_with_basics() {
    let input = quote! {
        impl SomeService {
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_lifetimes_and_generics() {
    let input = quote! {
        impl<'a, 'b, T, U> SomeService<'a, 'b, T, U>
        where T : Clone,
              U: Iterator<Item = u32> {
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_extends() {
    let args = quote! {
        extends = [ExtendedService1, ExtendedService2],
        //arg42 = "Hello, World!"
    };
    let input = quote! {
        impl SomeService {
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_events() {
    let args = quote! {
        events = SomeEvents,
    };
    let input = quote! {
        impl SomeService {
            pub fn do_this(&mut self) -> u32 {
                42
            }

            pub fn this(&self) -> bool {
                true
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_lifetimes_and_events() {
    let args = quote! {
        events = MyEvents,
    };
    let input = quote! {
        impl<'a, T> MyGenericEventsService<'a, T>
        where
            T: Clone,
        {
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_extends_and_lifetimes() {
    let args = quote! {
        extends = [base::BaseLifetime<'a>]
    };
    let input = quote! {
        impl<'a> ExtendedLifetime<'a> {
            pub fn extended_name(&self) -> String {
                "extended-name".to_string()
            }

            pub fn name(&self) -> String {
                "extended".to_string()
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_methods_with_lifetimes() {
    let input = quote! {
        impl ReferenceService {
            pub fn baked(&self) -> &'static str {
                "Static str!"
            }

            pub fn incr(&mut self) -> &'static ReferenceCount {
                unsafe {
                    COUNTER.0 += 1;
                    &*ptr::addr_of!(COUNTER)
                }
            }

            pub fn add_byte(&mut self, byte: u8) -> &'static [u8] {
                unsafe {
                    BYTES.push(byte);
                    &*ptr::addr_of!(BYTES)
                }
            }

            pub async fn first_byte<'a>(&self) -> Option<&'a u8> {
                unsafe { BYTES.first() }
            }

            pub async fn last_byte<'a>(&self) -> Option<&'a u8> {
                unsafe { BYTES.last() }
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_crate_path() {
    let args = quote!(crate = sails_rename,);
    let input = quote! {
        impl SomeService {
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}
