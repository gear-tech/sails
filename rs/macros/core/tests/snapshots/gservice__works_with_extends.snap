---
source: rs/macros/core/tests/gservice.rs
expression: result
---
impl SomeService {
    pub fn do_this(&mut self) -> u32 {
        42
    }
}
pub struct SomeServiceExposure<T> {
    message_id: sails_rs::MessageId,
    route: &'static [u8],
    #[cfg(not(target_arch = "wasm32"))]
    inner: Box<T>,
    #[cfg(not(target_arch = "wasm32"))]
    inner_ptr: *const T,
    #[cfg(target_arch = "wasm32")]
    inner: T,
    base: (
        <ExtendedService1 as sails_rs::gstd::services::Service>::Exposure,
        <ExtendedService2 as sails_rs::gstd::services::Service>::Exposure,
    ),
}
impl SomeServiceExposure<SomeService> {
    pub fn do_this(&mut self) -> u32 {
        let exposure_scope = sails_rs::gstd::services::ExposureCallScope::new(self);
        self.inner.do_this()
    }
    pub fn as_base_0(
        &self,
    ) -> &<ExtendedService1 as sails_rs::gstd::services::Service>::Exposure {
        &self.base.0
    }
    pub fn as_base_1(
        &self,
    ) -> &<ExtendedService2 as sails_rs::gstd::services::Service>::Exposure {
        &self.base.1
    }
    pub async fn try_handle(&mut self, input: &[u8]) -> Option<(Vec<u8>, u128)> {
        use sails_rs::gstd::InvocationIo;
        if let Ok(request) = some_service_meta::__DoThisParams::decode_params(input) {
            let result = self.do_this();
            let value = 0u128;
            let output = some_service_meta::__DoThisParams::encode_reply(&result);
            return Some((output, value));
        }
        if let Some((output, value)) = self.base.0.try_handle(input).await {
            return Some((output, value));
        }
        if let Some((output, value)) = self.base.1.try_handle(input).await {
            return Some((output, value));
        }
        None
    }
}
impl sails_rs::gstd::services::Exposure for SomeServiceExposure<SomeService> {
    fn message_id(&self) -> sails_rs::MessageId {
        self.message_id
    }
    fn route(&self) -> &'static [u8] {
        self.route
    }
}
impl sails_rs::gstd::services::Service for SomeService {
    type Exposure = SomeServiceExposure<Self>;
    fn expose(
        self,
        message_id: sails_rs::MessageId,
        route: &'static [u8],
    ) -> Self::Exposure {
        #[cfg(not(target_arch = "wasm32"))]
        let inner_box = Box::new(self);
        #[cfg(not(target_arch = "wasm32"))]
        let inner = inner_box.as_ref();
        #[cfg(target_arch = "wasm32")]
        let inner = &self;
        Self::Exposure {
            message_id,
            route,
            base: (
                <ExtendedService1 as Clone>::clone(
                        AsRef::<ExtendedService1>::as_ref(inner),
                    )
                    .expose(message_id, route),
                <ExtendedService2 as Clone>::clone(
                        AsRef::<ExtendedService2>::as_ref(inner),
                    )
                    .expose(message_id, route),
            ),
            #[cfg(not(target_arch = "wasm32"))]
            inner_ptr: inner_box.as_ref() as *const Self,
            #[cfg(not(target_arch = "wasm32"))]
            inner: inner_box,
            #[cfg(target_arch = "wasm32")]
            inner: self,
        }
    }
}
impl sails_rs::meta::ServiceMeta for SomeService {
    type CommandsMeta = some_service_meta::CommandsMeta;
    type QueriesMeta = some_service_meta::QueriesMeta;
    type EventsMeta = some_service_meta::EventsMeta;
    const BASE_SERVICES: &'static [sails_rs::meta::AnyServiceMetaFn] = &[
        sails_rs::meta::AnyServiceMeta::new::<ExtendedService1>,
        sails_rs::meta::AnyServiceMeta::new::<ExtendedService2>,
    ];
}
mod some_service_meta {
    use super::*;
    use sails_rs::{Decode, TypeInfo};
    #[derive(Decode, TypeInfo)]
    #[codec(crate = sails_rs::scale_codec)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub struct __DoThisParams {}
    impl sails_rs::gstd::InvocationIo for __DoThisParams {
        const ROUTE: &'static [u8] = &[24u8, 68u8, 111u8, 84u8, 104u8, 105u8, 115u8];
        type Params = Self;
    }
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum CommandsMeta {
        DoThis(__DoThisParams, u32),
    }
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum QueriesMeta {}
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum NoEvents {}
    pub type EventsMeta = NoEvents;
}
